src/lib/GenericPredictor.cc

```diff
@@ -23,25 +23,25 @@ GenericPredictor::GenericPredictor(ryml::ConstNodeRef config)
   btb_ =
       std::vector<std::pair<uint8_t, uint64_t>>(1 << btbBits_, {satCntVal, 0});
   // Alter globalHistoryLength_ value to better suit required format in update()
-  globalHistoryLength_ = (1 << globalHistoryLength_) - 1;
+  globalHistoryLength_ = (1 << (globalHistoryLength_ * 2)) - 1;
 }
 
 GenericPredictor::~GenericPredictor() {
   btb_.clear();
   ras_.clear();
   rasHistory_.clear();
+  ftq_.clear();
 }
 
 BranchPrediction GenericPredictor::predict(uint64_t address, BranchType type,
                                            int64_t knownOffset) {
-  // Get index via an XOR hash between the global history and the lower btbBits_
-  // bits of the instruction address
-  uint64_t hashedIndex = (address & ((1 << btbBits_) - 1)) ^ globalHistory_;
-  btbHistory_[address] = hashedIndex;
+  // Get index via an XOR hash between the global history and the instruction
+  // address. This hash is then ANDed to keep it within bounds of the btb
+  uint64_t hashedIndex =
+      ((address >> 2) ^ globalHistory_) & ((1 << btbBits_) - 1);
```

Comment:
Added

Reply:
@dANW34V3R -- you're correct that the language of my comment was off, I've updated it.  Re the 2-byte aligned instructions, I don't think that this makes this approach any less valid.  There is always going to be some aliasing in the BTB, but by ignoring the LSBs that may be zero in a lot of cases we maximise the information that is going into the hashes and therefore improve accuracy (in the generic predictor, this change alone improved misprediction rates modestly).  Indeed, any two instructions whose addresses share the same btbBits_ will have the same hash (of course assuming the same global history).

In relation to your second comment, there is no one correct way to do hashing, so so long as it was internally consistent before (which I think it was) it would have worked.  The new one should be marginally better though.

Reply:
Sorry for the long reply. The new comment is much better.

"There is always going to be some aliasing" - Yes, but some aliasing will have more of an effect than others. Branches that are adjacent in memory are more likely to be executed temporally close together (this claim needs some experimentation and backing though). Given that ~80% of instructions are compressed when using the C extension I think there is a claim that aliasing in this way could happen very often (again needs experimentation). With 4 byte aligned (always) instructions this is absolutely the correct thing to do, I'm just concerned about the impact when that isn't the case. The case being RV64 C which we already support. I think a small amount of experimentation of this case would be good here. If it is significant then this could either be parameterised or determined by querying the byte alignment from the architecture and setting accordingly. If not then just leave as is.

I'm not sure what you mean by "sharing btbBits_", this will be constant as it is a config parameter and only determines the size of the BTB and isn't instruction dependent. It is being used to create a mask to take only the correct number of LSB's from the produced hash (I think a tiny bit more explanation along these lines would help the comment as well) as to not index past the end of the BTB.

Of course there are many ways to create a hash. My thought was that with the previous calculation you can produce hashes which are well outside the addressable range of the array (but it won't complain it you do this) which will essentially return a random number. Previously you took the address and masked it to the range of the BTB length THEN you XOR'd it with the global history which could be up to 64 bits of 1s or 0s (unless I have missed something) which could introduce 1's into the upper bits of the index thus creating a larger number than the number of entries in the BTB. Not a big deal now that it has been fixed, but I am surprised that it worked before.

Reply:
I'm happy to do some experimentation: do you have any codes I could run that have been compiled in both compressed and non-compressed RISCV?  With the speculative global history updating that is now done, I'm not sure temporal proximity of instructions makes any difference as the global history will be updated between even programatically adjacent branch instructions and so the hash would be different even if they were to have the same address.

By sharing btbBits, I mean that if btbBits = 8, say, then any two instructions whose addresses have the same eight least-significant-bits will look the same to the hashing function (of course the hashes would likely be different unless they were both encountered when the global history was in the same state).  

Previously, the global history was kept to the size of the btb whenever it was updated so the logic was handled elsewhere.  Because we need longer global histories for the perceptrons we can't limit its size to the size of the btb, so this logic had to be moved to the hash generation.

Reply:
Thanks for answering all the queries. Comment looks good now - potentially will need an update talking about the fact that not all our supported ISA are 4 byte aligned. I will send over some binaries to test but that doesn't need to be part of this PR. Resolving